src/java/org/apache/cassandra/replication/ForwardedWriteRequest.java

```diff
@@ -0,0 +1,193 @@
+        Token token = mutation.key().getToken();
+        ReplicaPlan.ForWrite plan = ReplicaPlans.forWrite(keyspace, consistencyLevel, token, ReplicaPlans.writeNormal);
+
+        Replica leader = getLeader(plan);
+
+        // Add callbacks for replicas to respond directly to coordinator
+        Message<ForwardedWriteRequest> toLeader = Message.out(Verb.FORWARDING_WRITE, this);
+        for (Replica replica : plan.contacts())
+        {
+            logger.debug("Adding forwarding callback for response from {} id {}", replica.endpoint(), toLeader.id());
+            MessagingService.instance().callbacks.addWithExpiration(handler, toLeader, replica.endpoint());
+        }
+
+        MessagingService.instance().send(toLeader, leader.endpoint());
+    }
+
+    public static final Serializer serializer = new Serializer();
+
+    public static class Serializer implements IVersionedSerializer<ForwardedWriteRequest>
+    {
+        @Override
+        public void serialize(ForwardedWriteRequest request, DataOutputPlus out, int version) throws IOException
+        {
+            out.writeInt(request.verb.id);
+            Mutation.serializer.serialize(request.mutation, out, version);
+            out.writeInt(request.consistencyLevel.code);
+        }
+
+        @Override
+        public ForwardedWriteRequest deserialize(DataInputPlus in, int version) throws IOException
+        {
+            Verb verb = Verb.fromId(in.readInt());
+            Mutation mutation = Mutation.serializer.deserialize(in, version);
+            ConsistencyLevel consistencyLevel = ConsistencyLevel.fromCode(in.readInt());
+            return new ForwardedWriteRequest(verb, mutation, consistencyLevel, Dispatcher.RequestTime.forImmediateExecution());
+        }
+
+        @Override
+        public long serializedSize(ForwardedWriteRequest request, int version)
+        {
+            long size = 0;
+            size += TypeSizes.INT_SIZE;
+            size += Mutation.serializer.serializedSize(request.mutation, version);
+            size += TypeSizes.INT_SIZE;
+            return size;
+        }
+    }
+
+    public static final VerbHandler verbHandler = new VerbHandler();
+
+    public static class VerbHandler implements IVerbHandler<ForwardedWriteRequest>
+    {
+        @Override
+        public void doVerb(Message<ForwardedWriteRequest> incoming)
+        {
+            logger.debug("Received incoming ForwardedWriteRequest {} id {}", incoming, incoming.id());
+            Mutation mutation = incoming.payload.mutation;
+            assert mutation.id().isNone();
+            ConsistencyLevel consistencyLevel = incoming.payload.consistencyLevel;
+
+            Stage.MUTATION.submit(() -> {
+                new TrackedWriteRequest(DirectAcknowledge.toCoordinator(incoming.from(), incoming.id()))
+                        .perform(mutation, consistencyLevel, Dispatcher.RequestTime.forImmediateExecution());
+            });
```

Comment:
This shouldn't all run on MUTATION, and needs better error-handling; fix in-progress